### ISSUE TITLE
Add setting to remove really _all_ courses to cleaner_courses

### ISSUE THREAD

**Issue by abias** [catalyst/moodle-local_datacleaner#60]
Hi,

This problem report is targeted at the latest version of the plugin published on https://moodle.org/plugins/pluginversion.php?id=12918 and may be superseded by recent commits here in Github.

We would like to use local_datacleaner to remove really _all_ courses in one of our washing box instances.

On /admin/settings.php?section=cleaner_courses, we can only select the minimum age for a course to be deleted and the categories to be deleted. We have seen that, if we set minimum age = 0 and don't select any category, the plugin seems to interpret this to remove all courses, but unfortunately some courses remained undeleted.

Before we dive into debugging why some courses remain undeleted, I would like to ask if it's possible to add a setting which just removes really all courses in a washing box.

PS: Selecting all categories might be an intermediate solution, but I am afraid that this might also break as soon as new categories are created.

Thanks,
Alex

**Comment by brendanheywood** [comment]
That seems like a fine feature request as long as it defaulted off, but I'm curious what the use case is? If there is no content in a moodle then it's probably easier to start with an empty db and build it up rather than doing a lot of work to remove a lot of content.

Our typical use cases are that we want data, and it needs to be recent, but we don't want it to be large. So deleting old stuff is the natural solution.

If / when we do this I'd probably just clarify in the inline help that an age of 0 and no categories is exactly what you wanted and make it do that properly.

**Comment by abias** [comment]
> but I'm curious what the use case is? If there is no content in a moodle then it's probably easier to start with an empty db and build it up rather than doing a lot of work to remove a lot of content.

The use case is to have a Moodle playground instance which matches the codebase including plugins _and_ configuration of the production system. It will be used for evaluating plugins and custom developments in an environment which matches production as much as possible. But because of data security policies (and because there may be additional admins who are not allowed to see user data - see #63) we have to drop all existing user data from this playground instance.

> If / when we do this I'd probably just clarify in the inline help that an age of 0 and no categories is exactly what you wanted and make it do that properly.

So you are saying that age = 0 and categories = none should really remove all courses? In our tests, we have seen that there remain few courses in several categories. I would have to dive into debugging to identify these edge cases.

**Comment by brendanheywood** [comment]
> The use case is to have a Moodle playground instance which matches the codebase including plugins and configuration of the production system.

Right ok. Yeah I've wondered in the past for that use case whether a completely different approach would be better, more like starting from a fresh db and then building it up as needed. Some of our instances are truly massive and it's not great to constantly take one of them, clone it and then remove everything again which can take double digit hours. I've toyed with the idea of a simple plugin (which could be part of this plugin) which extracts all the config settings and exports it as a static config-include.php file.

So the whole process could be something like:

1) git clone fresh instance, setup db + filesystem etc (docker or whatever)
2) create minimal config.php file to get it running
3) add this to config.php

```php
if (file_exists(__DIR__.'/config-exported.php')) {
    require(__DIR__.'/config-exported.php');
}
```

On reflection I think it would be a great feature for inside this plugin and not a new tool, because it could take advantage of the matrix cleaner configuration which is already aware of what config should not be given out to dev instances. So perhaps this is a new tick box in the matrix cleaner plugin which runs a task each night which exports all the mdl_config and mdl_config_plugins tables, removes or swaps any values set in the matrix clean, and then produces a php file. To be safe I'd dump this into the moodle sitedata somewhere which can then be picked up and moved to where the dev's can access it. If you had admin access to prod you could also have a live export and download this file too. There are edges cases though with various plugins that don't store config in these tables though but I think that's manageable too, ie a selective export / import of just what is needed instead of an entire db dump. Dunno maybe that could get too hairy.

> So you are saying that age = 0 and categories = none should really remove all courses?

Yes I think that is a reasonable expectation so I'd treat this as a bug

**Comment by abias** [comment]
> Yeah I've wondered in the past for that use case whether a completely different approach would be better, more like starting from a fresh db and then building it up as needed. Some of our instances are truly massive and it's not great to constantly take one of them, clone it and then remove everything again which can take double digit hours. 
I agree with you that copying a whole instance to dump the biggest part of it later seems counterintuitive. However, going this way we can be sure that the test instance really has the configuration of the prod instance. And for us, copying the whole instance does not take more than one hour (yet).

> I've toyed with the idea of a simple plugin (which could be part of this plugin) which extracts all the config settings and exports it as a static config-include.php file.
This would be a nice idea. There is already a block which pretends to do this job on https://moodle.org/plugins/block_admin_presets, so maybe you don't have to reinvent the wheel completely. 

> Yes I think that is a reasonable expectation so I'd treat this as a bug
All right. Feel free to split the bug from the future improvement ideas discussed above to a separate ticket.
